core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java

```diff
@@ -123,189 +151,344 @@ public void unableToMonitorLockNode(Exception e) {
+              msgLoggedOnce = true;
+            }
+            LOG.debug("[{}] higher sequential node found: {}, deleting it", vmLockPrefix, child);
+            zooKeeper.delete(path + "/" + child);
+          }
+        }
+      }
+      final String pathForWatcher = lowestSequentialPath;
+
+      // Set a watcher on the lowest sequential node that we created, this handles the case
+      // where the node we created is deleted or if this client becomes disconnected.
+      LOG.debug("[{}] Setting watcher on {}", vmLockPrefix, pathForWatcher);
+      Watcher watcherForNodeWeCreated = new Watcher() {
 
         private void failedToAcquireLock() {
+          LOG.debug("[{}] Lock deleted before acquired, setting createdNodeName {} to null",
+              vmLockPrefix, createdNodeName);
           lw.failedToAcquireLock(new Exception("Lock deleted before acquired"));
-          asyncLock = null;
+          createdNodeName = null;
         }
 
         @Override
         public void process(WatchedEvent event) {
           synchronized (ZooLock.this) {
-            if (lock != null && event.getType() == EventType.NodeDeleted
-                && event.getPath().equals(path + "/" + lock)) {
+            if (lockNodeName != null && event.getType() == EventType.NodeDeleted
+                && event.getPath().equals(path + "/" + lockNodeName)) {
+              LOG.debug("[{}] {} was deleted", vmLockPrefix, lockNodeName);
               lostLock(LockLossReason.LOCK_DELETED);
-            } else if (asyncLock != null && event.getType() == EventType.NodeDeleted
-                && event.getPath().equals(path + "/" + asyncLock)) {
+            } else if (createdNodeName != null && event.getType() == EventType.NodeDeleted
+                && event.getPath().equals(path + "/" + createdNodeName)) {
+              LOG.debug("[{}] {} was deleted", vmLockPrefix, createdNodeName);
               failedToAcquireLock();
             } else if (event.getState() != KeeperState.Disconnected
-                && event.getState() != KeeperState.Expired && (lock != null || asyncLock != null)) {
-              log.debug("Unexpected event watching lock node {} {}", event, asyncLockPath);
+                && event.getState() != KeeperState.Expired
+                && (lockNodeName != null || createdNodeName != null)) {
+              LOG.debug("Unexpected event watching lock node {} {}", event, pathForWatcher);
               try {
-                Stat stat2 = zooKeeper.getStatus(asyncLockPath, this);
+                Stat stat2 = zooKeeper.getStatus(pathForWatcher, this);
                 if (stat2 == null) {
-                  if (lock != null)
+                  // if stat is null from the zookeeper.exists(path, Watcher) call, then we just
+                  // created a Watcher on a node that does not exist. Delete the watcher we just
+                  // created.
+                  zooKeeper.getZooKeeper().removeWatches(pathForWatcher, this, WatcherType.Any,
```

Comment:
Do you think calling removeWatches could be done in other places in ZooLock?  Wondering if there is a follow on issue.

Reply:
I think I caught all of the cases where a `Stat` is returned and it's `null`.